### ISSUE TITLE
Additional Audio Codec: FLAC

### ISSUE THREAD

**Issue by jprjr** [veovera/enhanced-rtmp#17]
Per https://github.com/veovera/enhanced-rtmp/issues/2#issuecomment-1533978134

Currently, the only way to deliver lossless audio over RTMP is via linear PCM, which has some drawbacks:

* Limited to 16-bit audio
* Max 2 channels
* Max sample rate of 44.1kHz

Adding FLAC would allow some flexibility for stream producers that want to transcode for multiple bandwidths without losing audio quality. Currently the best solution is to max out the AAC bitrate and hope there isn't too much quality loss on transcodes.

Plus, assuming FLAC would signal its own bit-depth, samplerate, and channel count (I know AAC signal its own samplerate, I can't recall if it also signals channel count) - you could stream incredibly high-quality audio.

As far as solutions shipping today that would benefit from FLAC support:

### OBS

Say I want to produce a web radio show. Most web radios are based on Icecast/shoutcast and a lot of the tooling around Icecast isn't great.

OBS is actually a pretty decent tool for producing radio-type content with builtin audio filters, support for multiple audio devices, scripting, hotkeys, and so on.

If OBS could output with FLAC over RTMP - a receiving server could take the single RTMP stream, and transcode to different formats (output to Icecast, HLS, and DASH) with significantly less quality loss.

### HLS/DASH repackagers

Similar to above - I know the various nginx RTMP add-ons support converting RTMP into HLS and DASH. Being able to accept a FLAC stream would allow for producing HLS, DASH, and even other RTMP outputs with multiple codecs for multiple devices.

Say I want to create a multivariant playlist in HLS with next-generation codecs like xHE-AAC. Having a single endpoint with a lossless codec would allow me to create a multivariant playlist with perfect sample alignment (since it's all from the same source). Currently I need to have my apps produce multiple streams at multiple bitrates and its hard to get everything aligned perfectly, or I have to just stream with AAC and re-encode/repackage server side and lose quality.

-----

Those are two off the top of my head, but it would really expand the possibilities for using RTMP as an intermediary/transport format. A user producing a web radio could stream in FLAC to a server over RTMP, which transcodes to formats like AAC, AAC-HE, xHE-AAC as HLS, and allow the listener to auto-fallback to codecs based on their bandwidth.

Long story short, the audio producer wouldn't need to worry if the final output/destination supports a particular codec - they would just stream in FLAC and have the ingest handle getting the right codecs out to the receivers. FLAC essentially allows nearly any other codecs to be supported via server-side transcoding.

**Comment by veovera** [comment]
this is a really nice writeup and I really appreciate the effort taken to explain why FLAC matters with specific real world scenarios. Because of your and previous suggestion FLAC is a codec we are considering.

**Comment by zenomt** [comment]
> Currently, the only way to deliver lossless audio over RTMP is via linear PCM

theoretically, HD-AAC (MPEG-4 Audio Scalable to Lossless/SLS aka ISO/IEC 14496-3 subpart 12), when coded over the lossy AAC Core Layer, should be backward compatible with AAC. its bitstream (including its `SLSSpecificConfig` structure which goes in the AAC `AudioSpecificConfig` block) could just go in today's RTMP AAC messages.

i suspect FLAC has wider tooling support though, especially free & open source.

**Comment by YellowOnion** [comment]
FLAC has comparably decent performance compared to the ISO standards or Apple lossless.

* Has a good balance between compression and complexity, unlike basically every other lossless codec.
* It is very cheap bit-stream to decode, some portable MP3 devices (before the iPhone was a thing) could get better battery life using FLAC compared to PCM or cheap lossy codecs like MP3 due to it reducing reads from flash media and CPU load.

* FLAC feels like the defacto standard lossless codec for music, it's available via Bandcamp, and on lots of players, and devices support it. FLAC has unofficial support in web browsers and because webm is a subset of MKV, which supports FLAC, you can also trick most browsers in to playing it.

* Apple Lossless does seem to be the only other codec that has some traction due to the Itunes store, The MPEG standards and DTS-HD and software seem to only target BluRay production environments.

https://wiki.hydrogenaud.io/index.php?title=Lossless_comparison

**Comment by jprjr** [comment]
> FLAC has unofficial support in web browsers

I think FLAC is officially supported in web browsers. I don't think you need to do any tricks to make it play. I've opened bugs with Mozilla and Chromium in the past, related to FLAC, and they both accepted and fixed them. I don't think they would have done that it if it was unofficial.

One other thing worth mentioning - the MPEG4 Registration Authority lists it as a codec, meaning it can be used in MP4 files, and makes it suitable for streaming via Media Source Extensions. 

I believe that means (if FLAC were supported): you could run an nginx server with the [nginx-http-flv-module](https://github.com/winshining/nginx-http-flv-module), push a FLAC stream via enhanced RTMP, pull in an FLV via HTTP, and play it back with Media Source Extensions via [mpegts.s](https://github.com/xqq/mpegts.js).

**Comment by veovera** [comment]
Yes, FLAC is supported by all major browsers. FYI https://caniuse.com/flac

**Comment by ctcwired** [comment]
Worth mentioning in my previous attempts to stream FLAC back in the Icecast days, digital silence could lead to no data being transmitted, and thus cause various stalls or timeouts in other muxers, or in clients tuned into the stream.

So for anyone implementing the encoding side of this, make sure the `--limit-min-bitrate` flag gets set.

Reference: https://github.com/xiph/flac/issues/90
